include/xtensor/xfunction.hpp

```diff
@@ -768,16 +801,16 @@ namespace xt
     template <class S>
     inline bool xfunction_base<F, R, CT...>::broadcast_shape(S& shape, bool reuse_cache) const
     {
-        if (reuse_cache && m_shape_computed)
+        if(reuse_cache)
         {
-            std::copy(m_shape.cbegin(), m_shape.cend(), shape.begin());
-            return m_shape_trivial;
+            //First call will compute the shape if this has not been done so yet.
+            std::copy(this->shape().cbegin(), this->m_shape.cend(), this->m_shape.begin());
```

Comment:
Never do that, you are not guaranteed which argument will be called first, and this is far less readable than the previous version.